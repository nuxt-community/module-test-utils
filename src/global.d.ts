import { NuxtTestContext } from './context'

declare global {
  namespace NodeJS {
    interface Global {
      $nuxtTestContext: NuxtTestContext;
    }
  }
}
