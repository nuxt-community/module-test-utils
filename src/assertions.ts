import { getNuxt } from './nuxt'

type ModuleContainerMethod = 'addPlugin' | 'addLayout' | 'addErrorLayout' | 'addServerMiddleware' | 'requireModule'

export function expectModuleToBeCalledWith (method: ModuleContainerMethod, ...args) {
  const _args = []
  for (const arg of args) {
    if (arg === undefined) {
      break
    }
    _args.push(arg)
  }

  expect(getNuxt().moduleContainer[method]).toBeCalledWith(..._args)
}
