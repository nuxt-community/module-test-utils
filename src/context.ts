import { resolve } from 'path'
import defu from 'defu'
import { NuxtConfig, NuxtOptions } from '@nuxt/types'
import type { Browser, LaunchOptions } from 'playwright'

let currentContext: NuxtTestContext

export function createContext (options: Partial<NuxtTestOptions>): NuxtTestContext {
  const _options: NuxtTestOptions = defu(options, {
    testDir: resolve(process.cwd(), 'test'),
    fixture: 'fixture',
    configFile: 'nuxt.config.js',
    setupTimeout: 60000,
    server: options.browser,
    build: options.browser || options.server,
    config: {},
    browserOptions: {
      type: 'chromium'
    }
  })

  return setContext({ options: _options })
}

export function getContext (): NuxtTestContext {
  if (!currentContext) {
    throw new Error('No context is available. (Forgot calling setup or createContext?)')
  }

  return currentContext
}

export function setContext (context: NuxtTestContext): NuxtTestContext {
  currentContext = context

  return currentContext
}

export interface NuxtTestOptions {
  testDir: string
  fixture: string
  configFile: string
  rootDir: string
  buildDir: string
  config: NuxtConfig

  build: boolean

  generate: boolean
  generateOptions: {
    build: boolean
    init: boolean
  }

  setupTimeout: number
  waitFor: number

  browser: boolean
  browserOptions: {
    type: 'chromium' | 'firefox' | 'webkit'
    launch: LaunchOptions
  }

  server: boolean
}

export interface NuxtTestContext {
  options: NuxtTestOptions

  nuxt?: {
    options: NuxtOptions
    listen: (port?: number) => any
    ready: () => any
    close: () => any
    moduleContainer: any
  }

  browser?: Browser

  url?: string

  builder?: {
    build: () => any
  }
}
