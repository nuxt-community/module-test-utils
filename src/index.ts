import preset from './preset'

export * from './assertions'
export * from './browser'
export * from './build'
export * from './context'
export * from './generate'
export * from './nuxt'
export * from './server'
export * from './setup'

export default preset
