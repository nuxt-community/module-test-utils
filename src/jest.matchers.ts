import { NuxtTestContext } from './types'

declare global {
  namespace jest {
    interface Matchers<R> {
      toHaveCalledNuxtAddPlugin(plugin: any): CustomMatcherResult
      toHaveCalledNuxtAddLayout(layout: any, name?: string): CustomMatcherResult
      toHaveCalledNuxtAddErrorLayout(errorLayout: any): CustomMatcherResult
      toHaveCalledNuxtAddServerMiddleware(middleware: any): CustomMatcherResult
      toHaveCalledNuxtRequireModule(moduleOpts: any): CustomMatcherResult
    }
  }
}

expect.extend({
  toHaveCalledNuxtAddPlugin (ctx: NuxtTestContext, plugin: any) {
    expect(ctx.nuxt.moduleContainer.addPlugin).toBeCalledWith(plugin)

    return { pass: true, message: () => '' }
  },

  toHaveCalledNuxtAddLayout (ctx: NuxtTestContext, layout: any, name?: string) {
    if (name) {
      expect(ctx.nuxt.moduleContainer.addLayout).toBeCalledWith(layout, name)
    } else {
      expect(ctx.nuxt.moduleContainer.addLayout).toBeCalledWith(layout)
    }

    return { pass: true, message: () => '' }
  },

  toHaveCalledNuxtAddErrorLayout (ctx: NuxtTestContext, errorLayout: any) {
    expect(ctx.nuxt.moduleContainer.addErrorLayout).toBeCalledWith(errorLayout)

    return { pass: true, message: () => '' }
  },

  toHaveCalledNuxtAddServerMiddleware (ctx: NuxtTestContext, middleware: any) {
    expect(ctx.nuxt.moduleContainer.addServerMiddleware).toBeCalledWith(middleware)

    return { pass: true, message: () => '' }
  },

  toHaveCalledNuxtRequireModule (ctx: NuxtTestContext, moduleOpts: any) {
    expect(ctx.nuxt.moduleContainer.requireModule).toBeCalledWith(moduleOpts)

    return { pass: true, message: () => '' }
  }
})
