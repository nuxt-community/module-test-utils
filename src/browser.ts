import { getContext } from './context'
import { url } from './server'

export async function createBrowser () {
  const ctx = getContext()
  const tib = await import('tib')
  ctx.puppeteer = await tib.createBrowser('puppeteer')
}

export async function createPage (path: string) {
  const ctx = getContext()
  const page = await ctx.puppeteer.page(url(path))
  return page
}
