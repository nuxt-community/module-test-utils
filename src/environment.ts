import NodeEnvironment from 'jest-environment-node'
import type { EnvironmentContext } from '@jest/environment'
import type { Config } from '@jest/types'
import { name } from '../package.json'
import { createContext, getContext, NuxtTestOptions } from './context'
import { setupNuxtTest, teardownNuxtTest } from './setup'

type NuxtTestEnvironment = Config.ProjectConfig & NuxtTestOptions

export class NuxtEnvironment extends NodeEnvironment {
  options: Partial<NuxtTestOptions>

  readonly pragmas: any
  readonly features: string[]

  constructor (config: NuxtTestEnvironment, context: EnvironmentContext) {
    super(config)
    this.options = config[name] || {}
    this.pragmas = context.docblockPragmas
    this.features = this.pragmas.features ? this.pragmas.features.split(' ') : []
  }

  parseFeatures () {
    if (this.features.includes('build')) {
      this.options.build = true
    }

    if (this.features.includes('server')) {
      this.options.server = true
    }

    if (this.features.includes('generate')) {
      this.options.generate = true
    }

    if (this.features.includes('browser')) {
      this.options.browser = true
    }

    if (this.features.includes('debug')) {
      this.options.browserOptions = {
        type: this.options?.browserOptions?.type || 'chromium',
        launch: {
          devtools: true,
          slowMo: 1000
        }
      }
    }
  }

  parseOptions () {
    if (this.pragmas.testDir) {
      this.options.testDir = this.pragmas.testDir
    }

    if (this.pragmas.fixture) {
      this.options.fixture = this.pragmas.fixture
    }

    if (this.pragmas.configFile) {
      this.options.configFile = this.pragmas.configFile
    }

    if (this.pragmas.rootDir) {
      this.options.rootDir = this.pragmas.rootDir
    }

    if (this.pragmas.waitFor) {
      this.options.waitFor = Number(this.pragmas.waitFor)
    }
  }

  async setup () {
    await super.setup()

    this.parseFeatures()
    this.parseOptions()

    await setupNuxtTest(createContext(this.options), false)

    this.global.$nuxtTestContext = getContext()
  }

  async teardown () {
    await teardownNuxtTest()

    this.global.$nuxtTestContext = undefined

    await super.teardown()
  }
}
