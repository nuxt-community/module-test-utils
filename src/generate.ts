import { getContext } from './context'
import { loadNuxtPackage } from './nuxt'
import { build } from './build'

export async function generate () {
  const ctx = getContext()
  const { Generator } = await loadNuxtPackage()

  if (!ctx.builder) {
    await build()
  }

  const generator = new Generator(ctx.nuxt, ctx.builder)

  await generator.generate({ build: false })
}
