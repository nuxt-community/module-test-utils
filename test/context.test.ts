import { createContext, getContext, NuxtTestContext } from '../src'

describe('context', () => {
  test('should be error if no context available', () => {
    expect(() => getContext()).toThrowError('No context is available. (Forgot calling setup or createContext?)')
  })

  test('default values from context', () => {
    const ctx: NuxtTestContext = createContext({ testDir: 'foo' })
    expect(ctx).toMatchSnapshot()
  })
})
