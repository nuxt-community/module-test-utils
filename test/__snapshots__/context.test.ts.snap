// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context default values from context 1`] = `
Object {
  "_id": 0,
  "browserString": "puppeteer",
  "build": undefined,
  "buildTimeout": 60000,
  "config": Object {},
  "configFile": "nuxt.config.js",
  "fixture": "fixture",
  "server": undefined,
  "testDir": "foo",
}
`;
